retro_data_structures/formats/mrea.py

```diff
@@ -499,45 +507,330 @@ def script_layers(self) -> Iterator[ScriptLayerHelper]:
+}
+
+
+class Area:
+    _mrea: Mrea = None
+    _strg: Strg = None
+
+    def __init__(self, raw: Container, index: int, parent_mlvl: Mlvl):
+        self._raw = raw
+        self.asset_manager = parent_mlvl.asset_manager
+        self._flags = parent_mlvl.layer_flags_for_area_index(index)
+        self._layer_names = parent_mlvl.layer_names_for_area_index(index)
+        self._index = index
+        self._parent_mlvl = parent_mlvl
+
+    @property
+    def id(self) -> int:
+        return self._raw.internal_area_id
+
+    @property
+    def index(self) -> int:
+        return self._index
+
+    @property
+    def name(self) -> str:
+        try:
+            return self.strg.strings[0]
+        except UnknownAssetId:
+            return "!!" + self.internal_name
+
+    @name.setter
+    def name(self, value):
+        self.strg.set_string(0, value)
+
+    @property
+    def internal_name(self) -> str:
+        return self._raw.get("internal_area_name", "Unknown")
+
+    @property
+    def strg(self) -> Strg:
+        if self._strg is None:
+            self._strg = self.asset_manager.get_file(self._raw.area_name_id, type_hint=Strg)
+        return self._strg
+
+    @property
+    def mrea(self) -> Mrea:
+        if self._mrea is None:
+            self._mrea = self.asset_manager.get_file(self.mrea_asset_id, type_hint=Mrea)
+        return self._mrea
+
+    @property
+    def mrea_asset_id(self) -> int:
+        return self._raw.area_mrea_id
+
+    @property
+    def layers(self) -> Iterator[ScriptLayer]:
+        for layer in self.mrea.script_layers:
+            yield layer.with_parent(self)
+
     @property
-    def all_instances(self) -> Iterator[ScriptInstanceHelper]:
-        for layer in self.script_layers:
+    def generated_objects_layer(self) -> ScriptLayer:
+        return self.mrea.generated_objects_layer.with_parent(self)
+
+    def get_layer(self, name: str) -> ScriptLayer:
+        return next(layer for layer in self.layers if layer.name == name)
+
+    def add_layer(self, name: str, active: bool = True) -> ScriptLayer:
+        index = len(self._layer_names)
+        self._layer_names.append(name)
+        self._flags.append(active)
+        raw = new_layer(index, self.asset_manager.target_game)
+        self.mrea._raw.sections.script_layer_section.append(raw)
+        return self.get_layer(name)
+
+    @property
+    def next_instance_id(self) -> int:
+        ids = [instance.id.instance for instance in self.all_instances]
+        return next(i for i in itertools.count() if i not in ids)
+
+    @property
+    def all_instances(self) -> Iterator[ScriptInstance]:
+        for layer in self.layers:
             yield from layer.instances
+        yield from self.generated_objects_layer.instances
+
+    def get_instance(self, instance_id: InstanceRef) -> ScriptInstance:
+        instance_id = resolve_instance_ref(instance_id)
+        layer = self.get_layer_for_instance(instance_id)
+
+        if layer.has_instance(instance_id):
```

Comment:
I do not enjoy this `has_instance` -> `get_instance`, particularly because `has_instance` is just a `get_instance` with try except, and it iterates over all instances.